### ISSUE TITLE
Returning ret.$observable sometimes behaving unpredictably, breaking resolvers

### ISSUE THREAD

**Issue by dbettini** [troyanskiy/ngx-resource#99]
The returned value's $observable property doesn't behave like it should in some cases. Not sure what is the reason for this, maybe someone more well versed in observables can explain. 

In short, it started when I added two [resolvers](https://angular.io/docs/ts/latest/api/router/index/Resolve-interface.html) on the same route, which use ngx-resource to fetch data, by returning this.resource.getData().$observable. Sometimes, data of one of them would be missing in route data. The only way to fix it was ~~either~~ calling toPromise()~~, or returning Observable.of(this.resource.getData()), or just returning this.resource.getData(), since resolvers will wrap it in observables anyway.~~
**EDIT:** since ngx-resource returns the empty model before the request finished, I was wrong in thinking it works that way. It only works with toPromise().

The full explanation for what I did and found out can be found [here](https://github.com/angular/angular/issues/13946#issuecomment-292357833). 

Basically, $observable is behaving a bit different than a regular observable in some cases... For example, if you try calling $observable.first(), it throws "no elements in sequence" error. This also makes it unsafe to return $observable from resolvers, and makes you fallback to promises.

**Comment by troyanskiy** [comment]
Hello @dbettini I'm sorry for delayed reply.
Probably I'm `cooking` Observables wrong, I'm not a master of `Observables`.

What I'm doing there.
1. Creates kind of deferred Observable and assign it to `$observable` [ResourceAction.ts#L74](https://github.com/troyanskiy/ng2-resource-rest/blob/master/src/ResourceAction.ts#L74)
1. Creating the request object by resolving some promises [ResourceAction.ts#L96](https://github.com/troyanskiy/ng2-resource-rest/blob/master/src/ResourceAction.ts#L96)
1. Create an request with angular http [ResourceAction.ts#L287](https://github.com/troyanskiy/ng2-resource-rest/blob/master/src/ResourceAction.ts#L287) and subscribe to it is it's not lazy request [ResourceAction.ts#L287](https://github.com/troyanskiy/ng2-resource-rest/blob/master/src/ResourceAction.ts#L287)
1. Same time it creates `mainObservable` which you actually receives in `$observable`
1. After creating the `mainObservable` i'm releasing deferred Observable [ResourceAction.ts#L356](https://github.com/troyanskiy/ng2-resource-rest/blob/master/src/ResourceAction.ts#L356)
1. After receiving the response from server i'm forwarding the data to `mainObservable` [ResourceAction.ts#L331](https://github.com/troyanskiy/ng2-resource-rest/blob/master/src/ResourceAction.ts#L331), [ResourceAction.ts#L334](https://github.com/troyanskiy/ng2-resource-rest/blob/master/src/ResourceAction.ts#L334) and [ResourceAction.ts#L337](https://github.com/troyanskiy/ng2-resource-rest/blob/master/src/ResourceAction.ts#L337)

Probably I'm doing something wrong here, I will need to dig more deep into Observables.
If you have any ideas how it can be fixed, please share with me.

**Comment by dbettini** [comment]
Hello, check this [JSBin](http://jsbin.com/mokaxoqeho/1/edit?js,console).
```
// build a source from an empty array
const source = Rx.Observable.from([]);

//subscribing emits done
const subscribe = source.subscribe(
  
//subscribing throws EmptyError
//const subscribe = source.first().subscribe(
  
  val => console.log(val),
  err=>console.log(err.constructor),
  ()=>console.log("done")
);
```
Subscribing to source, which is made from an empty array, just emits "done". Subscribing to source.first() throws the "no elements in sequence" error.

The $observable you return is empty so it isn't possible to immediately call first() on it, so it throws the "no elements in sequence" error. Maybe resolvers return an empty value because it emits done before the request finishes. Not sure why one always succeeds, maybe because angular waits to have at least one value before resolving the route...
Like I said, not well versed with observables either, but this is the only thing that comes to mind...

**Comment by tarlepp** [comment]
Yep, I have this same issue within my current application.

**Comment by troyanskiy** [comment]
Hello all.

I've released some kind of beta of new library which is called `rest-core` + `rest-ngx`.
Please check them [`rest-core`](https://www.npmjs.com/package/rest-core) and [`rest-ngx`](https://www.npmjs.com/package/rest-ngx).

It works the way like `ngx-resource` but has a lot of breaking changes.
Something was removed or simplified.

I've migrated all my projects to the new library.
If you wish to switch to HttpClient or use some other http handlers like `fetch` try to migrate your projects to the lib.

Bugs and help requests are welcome in corresponding repos.

Thanks!

PS: Since `ngx-resource` is no longer supported by me, closing the issue.
